spec/02-integration/05-proxy/09-balancer_spec.lua

```diff
@@ -1331,37 +1331,110 @@ for _, strategy in helpers.each_strategy() do
+
+                local upstream_name = add_upstream({
+                  hash_on = "header",
+                  hash_on_header = "hashme",
+                })
+                local port1 = add_target(upstream_name, localhost)
+                local port2 = add_target(upstream_name, localhost)
+                local api_host = add_api(upstream_name)
+
+                -- setup target servers
+                local server1 = http_server(localhost, port1, { requests })
+                local server2 = http_server(localhost, port2, { requests })
+
+                -- Go hit them with our test requests
+                local oks = client_requests(requests, {
+                  ["Host"] = api_host,
+                  ["hashme"] = "just a value",
+                })
+                assert.are.equal(requests, oks)
+
+                -- collect server results; hitcount
+                -- one should get all the hits, the other 0
+                local _, count1 = server1:done()
+                local _, count2 = server2:done()
+
+                -- verify
+                assert(count1 == 0 or count1 == requests, "counts should either get 0 or ALL hits")
+                assert(count2 == 0 or count2 == requests, "counts should either get 0 or ALL hits")
+                assert(count1 + count2 == requests)
+              end)
+
+              describe("hashing on cookie", function()
+                it("when the cookie is set", function()
+                  local requests = SLOTS * 2 -- go round the balancer twice
+
+                  local upstream_name = add_upstream({
+                    hash_on = "cookie",
+                    hash_on_cookie = "hashme",
+                  })
+                  local port1 = add_target(upstream_name, localhost)
+                  local port2 = add_target(upstream_name, localhost)
+                  local api_host = add_api(upstream_name)
+
+                  -- setup target servers
+                  local server1 = http_server(localhost, port1, { requests })
+                  local server2 = http_server(localhost, port2, { requests })
+
+                  -- Go hit them with our test requests
+                  local oks = client_requests(requests, {
+                    ["Host"] = api_host,
+                    ["Cookie"] = "hashme=just-a-value",
+                  })
+                  assert.are.equal(requests, oks)
+
+                  -- collect server results; hitcount
+                  -- one should get all the hits, the other 0
+                  local _, count1 = server1:done()
+                  local _, count2 = server2:done()
+
+                  -- verify
+                  assert(count1 == 0 or count1 == requests, "counts should either get 0 or ALL hits")
+                  assert(count2 == 0 or count2 == requests, "counts should either get 0 or ALL hits")
+                  assert(count1 + count2 == requests)
+                end)
+
+                it("when the cookie is not set", function()
```

Comment:
There's a lot of redundancy between these two new examples. Not sure we really need to test when the cookie is already set, if we're testing when it's not set. Thoughts?